### PR TITLE
Fix bugs found in Network.Analytics RP

### PR DESCRIPTION
**Pull request by MaryGao**
Bugfixes found in Network.Analytics RP
- Only adopt getEffectiveModelType for anonymous model for modular and RLC
- Enable `readonly` when the visibility is only read in modular

fixes https://github.com/Azure/autorest.typescript/issues/2309 https://github.com/Azure/autorest.typescript/issues/2310  https://github.com/Azure/autorest.typescript/issues/2311

### CODE REVIEW

File: packages/typespec-test/test/anomalyDetector/generated/typespec-ts/review/ai-anomaly-detector.api.md [hunk 1]
```diff
@@ -359,7 +359,7 @@ export interface ListMultivariateModelsDefaultHeaders {
 // @public (undocumented)
 export interface ListMultivariateModelsDefaultResponse extends HttpResponse {
     // (undocumented)
-    body: MultivariateErrorResponseOutput;
+    body: MultivariateResponseErrorOutput;
```
Comment (qiaozha) [inline packages/typespec-test/test/anomalyDetector/generated/typespec-ts/review/ai-anomaly-detector.api.md:362]: Can you check if we still keep the original `MultivariateErrorResponseOutput` model ? I don't see where it's being deleted ?

File: packages/typespec-ts/test/modularUnit/modelsGenerator.spec.ts [hunk 2]
```diff
@@ -1954,3 +1952,66 @@ describe("inheritance & polymorphism", () => {
...
 });
+
+describe("`is`", () => {
+  it("should generate correct name and properties if A is B<Template>", async () => {
+    const modelFile = await emitModularModelsFromTypeSpec(`
+    model B<Parameter> {
+      prop1: string;
+      prop2: Parameter;
+    }
+    model A is B<string> {
```
Comment (qiaozha) [inline packages/typespec-ts/test/modularUnit/modelsGenerator.spec.ts:1963]: Is there a case for `extends` ?

**Review by qiaozha — APPROVED** [review]
LGTM, some minor comments

File: packages/typespec-ts/test/modularUnit/modelsGenerator.spec.ts (hunk 2 above)
Comment (MaryGao) [reply packages/typespec-ts/test/modularUnit/modelsGenerator.spec.ts:1963]: added.

File: packages/typespec-ts/src/modular/buildCodeModel.ts [hunk 3]
```diff
@@ -252,20 +252,25 @@ function getEffectiveSchemaType(program: Program, type: Model | Union): Model {
...
+      nonNullOptions.length === 1 &&
+      nonNullOptions[0]?.kind === "Model" &&
+      nonNullOptions[0]?.name === ""
+    ) {
       effective = getEffectiveModelType(program, nonNullOptions[0]);
     }
     return type as any;
-  } else {
+  } else if (type.name === "") {
     effective = getEffectiveModelType(program, type, isSchemaProperty);
```
Comment (MaryGao) [inline packages/typespec-ts/src/modular/buildCodeModel.ts:270]: Only apply getEffectiveModelType for anonymous model due to https://github.com/microsoft/typespec/issues/2991#issuecomment-1983609082

**Pull request merged**